File: primos.py
from math import sqrt

def listar_primos(number: int) -> list[int]:
    """
        Função que encontra todos os números primos entre 1 e o número especificado (N).
    
        Args:
            number (int): O número até o qual os primos serão encontrados. Deve ser um número inteiro positivo.

        Returns:
            list[int]: Uma lista contendo todos os números primos entre 1 e o número especificado (inclusive).
        
        Exemplo:
            >>> encontrar_primos_entre_1_e_n(10)
            [2, 3, 5, 7]
    """
    # Validações iniciais do número
    try:
        validate_input(number)
    except (TypeError, ValueError) as e:
        return e
    
    # Inicializa a lista de primos vazia
    primos = []

    # Percorremos todos os números de 2 até o número especificado e retornamos os valores primos
    for i in range(2, number + 1):
        if is_prime(i):
            primos.append(i)
    return primos


def listar_primos_rec(number: int, primos: list[int]=[], current: int=2)-> list[int]:
    """
        Função recursiva que encontra todos os números primos entre 1 e o número especificado (N).

        Args:
            number (int): O número até o qual os primos serão encontrados.
            current (int): O número atual sendo verificado, sendo 2 o primeiro número a ser verificado.
            primos (list[int]): A lista de números primos encontrados até o momento, inicialmente vazia.
        
        Returns:
            list[int]: A lista de números primos encontrados até o número especificado.
    """
    # Validação inicial, com o valor padrão para o número atual
    if current == 2:
        try:
            validate_input(number)
        except (ValueError, TypeError) as e:
            return e
    
    # Se chegar a um valor maior que o número passado, retornamos a lista de valores primos encontrados
    if current > number:
        return primos
    
    # Se o valor atual for primo, adicionamos a lista de primos
    if is_prime(current):
        primos.append(current)

    # Chamada recursiva para o próximo número
    return listar_primos_rec(number, primos, current + 1)


def is_prime(number):
    """
        Verifica se um número é primo

        Args:
            number (int): O número a ser verificado.

        Returns:
            Bool: True se o número é primo, False caso contrário
    """
    if number == 2:
        return True
    # Verifica se o número é par, para ajudar na otimizar da verificação, já que números pares maiores que 2 não são primos (são divisiveis por 2)
    if number <= 1 or number % 2 == 0:
        return False
    
    # Percorre os números ímpares de forma decrescente, entre a raiz quadrada do número e 2, para verificar se o número é divisível por algum deles.
    for i in range(int(sqrt(number)),2,-2):
        if number % i == 0:
            return False
    return True

def validate_input(number: int):
    """
        Valida o número passado como entrada, garantindo que seja um número inteiro positivo.
        Retorna um erro caso a entrada seja inválida.
    """
    if not isinstance(number, int):
        raise TypeError("O número deve ser um número inteiro.")
    if number < 1:
        raise ValueError("O número deve ser um número inteiro positivo.")
    
if __name__ == "__main__":
    # Valores inválidos
    print(listar_primos(0)) 
    print(listar_primos('a'))

    # Realizando testes
    print(listar_primos(1)) 
    print(listar_primos_rec(1))
    print(listar_primos_rec(100))
    print(listar_primos(100))
        
    

File: primos.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Primos {

    /**
     * Lista todos os números primos entre 1 e o número especificado (incluso).
     *
     * @param number O número até o qual os primos serão encontrados (deve ser positivo).
     * @return Uma lista contendo todos os números primos entre 1 e o número especificado.
     */
    public static List<Long> listarPrimos(long number) {
        // Retorna lista vazia se o número for menor que 1
        if (number < 1) {
            return new ArrayList<>();
        }
        
        // Inicia a lista de primos vazia, e itera por todos os valores de 2 a n procurando por primos
        List<Long> primos = new ArrayList<>();
        for (long i = 2; i <= number; i++) {
            if (isPrime(i)) {
                primos.add(i);
            }
        }
        return primos;
    }

    /**
     * Função que inicializa as variáveis necessárias e chama a função recursiva que encontra os primos menores ou iguais a n
     *
     * @param number O número até o qual os primos serão encontrados (deve ser positivo).
     * @return Uma lista contendo todos os números primos entre 1 e o número especificado.
     */
    public static List<Long> iniciarListagemPrimosRec(long number) {
        // Inicializa a lista de primos vazia e com o numero atual sendo igual a 2
        return listarPrimosRec(number, new ArrayList<>(), 2); 
    }

    /**
     * Lista todos os números primos entre 1 e o número especificado (incluso) de forma recursiva.
     *
     * @param number O número até o qual os primos serão encontrados (deve ser positivo).
     * @param primos A lista de números primos encontrados até agora.
     * @param current O número atual sendo verificado (começa em 2).
     * @return Uma lista contendo todos os números primos entre 1 e o número especificado.
     */
    public static List<Long> listarPrimosRec(long number, List<Long> primos, long current) {
        // Verifica se o número atual ultrapassou o limite
        if (current > number) {
            return primos;
        }
    
        // Se o número é primo, adiciona à lista
        if (isPrime(current)) {
            primos.add(current);
        }
    
        // Chama recursivamente para o próximo número
        return listarPrimosRec(number, primos, current + 1);
    }

    /**
     * Verifica se um número é primo.
     *
     * @param number O número a ser verificado.
     * @return `true` se o número for primo, caso contrário `false`.
     */
    public static boolean isPrime(long number) {
        // Caso o número seja par e não for 2 retorna False (2 é o único par primo)
        if (number == 2) return true;
        if (number <= 1 || number % 2 == 0) return false;
        
        // Caso o número seja impar, verifica se  algum valor divide ele
        long limit = (long) Math.sqrt(number);
        for (long i = 3; i <= limit; i += 2) {
            if (number % i == 0) return false;
        }
        return true;
    }
    

    public static void main(String[] args) {
        try (Scanner scanner = new Scanner(System.in)) {
            // Solicita ao usuário o valor até o qual ele deseja encontrar os números primos.
            System.out.println("Digite um número para listar os primos até esse valor:");
            long numero = scanner.nextLong();

            // Testando valores válidos com entrada do usuário
            System.out.println("Primos até " + numero + " (versão iterativa):");
            System.out.println(listarPrimos(numero)); 

            // Testando a função recursiva com inicialização separada
            System.out.println("Primos até " + numero + " (versão recursiva):");
            System.out.println(iniciarListagemPrimosRec(numero));
        }  catch (Exception e) {
            System.out.println("Erro ao ler entrada do usuário: " + e.getMessage());
        }
    }
}
